java-checks/src/main/java/org/sonar/java/checks/helpers/HardcodedStringExpressionChecker.java

```diff
@@ -0,0 +1,207 @@
+      case TYPE_CAST:
+        TypeCastTree typeCast = (TypeCastTree) arg;
+        return isExpressionDerivedFromPlainText(typeCast.expression(), secondaryLocations, visited);
+      case BOOLEAN_LITERAL:
+      case CHAR_LITERAL:
+      case DOUBLE_LITERAL:
+      case FLOAT_LITERAL:
+      case INT_LITERAL:
+      case LONG_LITERAL:
+        return true;
+      default:
+        if (arg instanceof BinaryExpressionTree) {
+          BinaryExpressionTree binaryExpression = (BinaryExpressionTree) arg;
+          return isDerivedFromPlainText(binaryExpression, secondaryLocations, visited);
+        }
+        return false;
+    }
+  }
+
+  private static boolean isDerivedFromPlainText(BinaryExpressionTree binaryExpression, List<JavaFileScannerContext.Location> secondaryLocations,
+    Set<Symbol> visited) {
+    return isExpressionDerivedFromPlainText(binaryExpression.rightOperand(), secondaryLocations, visited) &&
+      isExpressionDerivedFromPlainText(binaryExpression.leftOperand(), secondaryLocations, visited);
+  }
+
+  private static boolean isDerivedFromPlainText(IdentifierTree identifier, List<JavaFileScannerContext.Location> secondaryLocations,
+    Set<Symbol> visited) {
+    Symbol symbol = identifier.symbol();
+    boolean firstVisit = visited.add(symbol);
+    if (!firstVisit || !symbol.isVariableSymbol() || JUtils.isParameter(symbol) || isNonFinalField(symbol)) {
+      return false;
+    }
+    VariableTree variable = (VariableTree) symbol.declaration();
+    if (variable == null) {
+      return JUtils.constantValue((Symbol.VariableSymbol) symbol).isPresent();
+    }
+
+    List<ExpressionTree> assignments = ExpressionsHelper.getIdentifierAssignments(identifier);
+
+    boolean identifierIsDerivedFromPlainText = !assignments.isEmpty() &&
+      assignments.stream()
+        .allMatch(expression -> isExpressionDerivedFromPlainText(expression, secondaryLocations, visited));
+
+    if (identifierIsDerivedFromPlainText) {
+      secondaryLocations.add(new JavaFileScannerContext.Location("", variable));
+      return true;
+    }
+    return false;
+  }
+
+  private static boolean isNonFinalField(Symbol symbol) {
+    return symbol.isVariableSymbol() && symbol.owner().isTypeSymbol() && !symbol.isFinal();
+  }
+
+  private static boolean isDerivedFromPlainText(NewArrayTree invocation, List<JavaFileScannerContext.Location> secondaryLocations,
+    Set<Symbol> visited) {
+    ListTree<ExpressionTree> initializers = invocation.initializers();
+    return !initializers.isEmpty() && initializers.stream()
+      .allMatch(expression -> isExpressionDerivedFromPlainText(expression, secondaryLocations, visited));
+  }
+
+  /**
+   * When a constructor is found during the evaluation, we check if it matches with one of the specified constructors
+   * we passed to the Evaluator. If it does match, we will recursively evaluate the parameter targeted by the associated
+   * index in the supportedConstructors map.
+   */
```

Comment:
You don't need a javadoc comment, the ordinary one should be enough.
Also, it looks like there is nothing special in this comment that you can't understand from the code. Do you think it's still necessary?